init.zsh

```diff
@@ -1,23 +1,178 @@
+  else
+    antigen-hs-sandbox-cabal-check
+  fi
+}
+
+antigen-hs-repeat () {
+  if ! eval $1 ; then
+    antigen-hs-ask "$2 finish without success. Would you like try again?" "$0 $*"
+  fi
+}
+
+antigen-hs-init () {
+  antigen-hs-ask "Use sandbox for haskell dependencies?" antigen-hs-init-sandbox antigen-hs-cabal-global
+}
+
+antigen-hs-setup () {
+  antigen-hs-repeat antigen-hs-init "Setup"
+  antigen-hs-repeat antigen-hs-compile "Compilation"
+}
+
+antigen-hs-source () {
   local FILE_TO_SOURCE="$ANTIGEN_HS_OUT/antigen-hs.zsh"
-  if [[ -f $FILE_TO_SOURCE ]]
-  then
+
+  if [[ -f $FILE_TO_SOURCE ]] ; then
     source $FILE_TO_SOURCE
   else
-    echo "Didn't find file $FILE_TO_SOURCE"
-    echo "Try running antigen-hs-compile"
+    echo $fg[red]"Didn't find file $FILE_TO_SOURCE"$reset_color
+    return 1
+  fi
+}
+
+antigen-hs-init-source () {
+  antigen-hs-source
+  local ANTIGEN_HS_SOURCE_RESULT=$?
+
+  if [[ "$ANTIGEN_HS_SOURCE_RESULT" == 0 ]] ; then
+    if [[ -z $ANTIGEN_HS_SANDBOX ]]; then
+
+      if [[ -d "$ANTIGEN_HS_HOME"/.stack-work ]] ; then
+        ANTIGEN_HS_SANDBOX=$ANTIGEN_HS_SANDBOX_STACK
+      elif [[ -d "$ANTIGEN_HS_HOME"/.cabal-sandbox ]] ; then
+        ANTIGEN_HS_SANDBOX=$ANTIGEN_HS_SANDBOX_CABAL
+      else
+        ANTIGEN_HS_SANDBOX=$ANTIGEN_HS_SANDBOX_EMPTY
+      fi
+
+    fi
+  else
+    return "$ANTIGEN_HS_SOURCE_RESULT"
+  fi
+}
+
+antigen-hs-compile () {
+  local ANTIGEN_HS_COMPILE_CMD='runghc -i"$ANTIGEN_HS_HOME/" -- "$ANTIGEN_HS_MY"'
+
+  case $ANTIGEN_HS_SANDBOX in
+    $ANTIGEN_HS_SANDBOX_STACK )
+      local ANTIGEN_HS_COMPILE_CMD='STACK_YAML="$ANTIGEN_HS_HOME"/stack.yaml stack exec -- '"$ANTIGEN_HS_COMPILE_CMD"
+      unfunction antigen-hs-sandbox-cabal-check 2>/dev/null
+      unfunction antigen-hs-sandbox-cabal 2>/dev/null
+      unfunction antigen-hs-cabal 2>/dev/null
+      ;;
+    $ANTIGEN_HS_SANDBOX_CABAL )
+      local ANTIGEN_HS_COMPILE_CMD='CABAL_SANDBOX_CONFIG="$ANTIGEN_HS_HOME"/cabal.sandbox.config cabal exec -- '"$ANTIGEN_HS_COMPILE_CMD"
+      unfunction antigen-hs-sandbox-stack 2>/dev/null
+      ;;
+    * )
+      unfunction antigen-hs-sandbox-cabal-check 2>/dev/null
+      unfunction antigen-hs-sandbox-cabal 2>/dev/null
+      unfunction antigen-hs-sandbox-stack 2>/dev/null
+      unfunction antigen-hs-sandbox 2>/dev/null
+      ;;
+  esac
+  eval "$ANTIGEN_HS_COMPILE_CMD"
+  local ANTIGEN_HS_COMPILE_RESULT=$?
+
+  if [[ "$ANTIGEN_HS_COMPILE_RESULT" == 0 ]] ; then
+    antigen-hs-source
+  else
+    return "$ANTIGEN_HS_COMPILE_RESULT"
+  fi
+}
+
+() {
+  autoload -U colors && colors
+  if ! antigen-hs-init-source ; then
+    antigen-hs-ask "Try to setup?" antigen-hs-setup
   fi
+  unfunction antigen-hs-init-source
+  unfunction antigen-hs-setup
+  unfunction antigen-hs-repeat
+  unfunction antigen-hs-init
+  unfunction antigen-hs-cabal-global
+  unfunction antigen-hs-init-sandbox
```

Comment:
In overall for this file. I think it's hard to see what's public and what's private. Can you change the internal global variables from `$ANTIGEN_HS_BLAH` to `$_ANTIGEN_HS_BLAH`?